utils/tx/bank_send.go

```diff
@@ -0,0 +1,169 @@
+package tx
+
+import (
+	"errors"
+	"math/big"
+	"strconv"
+
+	sdkmath "cosmossdk.io/math"
+	tmbytes "github.com/cometbft/cometbft/libs/bytes"
+	tmtypes "github.com/cometbft/cometbft/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/haqq-network/haqq/utils"
+	evmtypes "github.com/haqq-network/haqq/x/evm/types"
+)
+
+func BankSendTxAsEthereumTx(tx sdk.Tx) (*evmtypes.MsgEthereumTx, error) {
+	if tx == nil {
+		return nil, errors.New("tx is nil")
+	}
+
+	if len(tx.GetMsgs()) == 0 {
+		return nil, errors.New("tx has no messages")
+	}
+
+	if len(tx.GetMsgs()) > 1 {
+		return nil, errors.New("tx has more than one message")
+	}
+
+	msg := tx.GetMsgs()[0]
+	sendMsg, ok := msg.(*banktypes.MsgSend)
+	if !ok {
+		return nil, errors.New("tx message is not a bank.MsgSend")
+	}
+
+	// Get gas if set
+	dTx, ok := tx.(sdk.FeeTx)
+	gasLimit := uint64(0)
+	gasPrice := sdkmath.ZeroInt()
+	if ok {
+		gasLimit = dTx.GetGas()
+		if gasLimit > 0 {
+			gasPrice = sdkmath.NewIntFromBigInt(dTx.GetFee().AmountOf(utils.BaseDenom).BigInt()).Quo(sdkmath.NewIntFromUint64(gasLimit))
+		}
+	}
+
+	sigTx, ok := tx.(authsigning.SigVerifiableTx)
+	if !ok {
+		return nil, errors.New("transaction without signatures")
+	}
+
+	sigs, err := sigTx.GetSignaturesV2()
+	if err != nil {
+		return nil, errors.New("failed to get transaction signatures")
+	}
+
+	switch {
+	case len(sigTx.GetSigners()) == 1:
+		if len(sigs) != 1 {
+			return nil, errors.New("transaction with one signer has more than one signature")
+		}
```

Comment:
error message could be clearer here:


```suggestion
		if len(sigs) != 1 {
			return nil, fmt.Errorf("expected transaction with 1 signer; got %d", len(sigs))
		}
```